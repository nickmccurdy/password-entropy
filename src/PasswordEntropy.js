import CustomOptions from './CustomOptions'
import DicewareOptions from './DicewareOptions'
import GenericOptions from './GenericOptions'
import handleChange from './handleChange'
import React, { PureComponent } from 'react'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'
import 'react-tabs/style/react-tabs.css'

export default class PasswordEntropy extends PureComponent {
  static optionsComponents = [
    DicewareOptions,
    CustomOptions,
    GenericOptions
  ]
  static tabs = PasswordEntropy.optionsComponents.map(({ name }) => (
    <Tab key={name}>{name.replace(/Options$/, '')}</Tab>
  ))

  handleChange = handleChange.bind(this)
  handlePossibleItemsChange = possibleItems => this.setState({ possibleItems })
  state = {
    length: 6
  }
  tabPanels = PasswordEntropy.optionsComponents.map(Component => (
    <TabPanel key={Component.name}>
      <Component onChange={this.handlePossibleItemsChange}/>
    </TabPanel>
  ))

  render () {
    const possiblePasswords = (this.state.possibleItems || 1) ** this.state.length
    const approximatePrefix = possiblePasswords > Number.MAX_SAFE_INTEGER && '~ '

    return (
      <form>
        <h1>Password Entropy</h1>

        <label>
          <h2>Length</h2>
          <input name="length" value={this.state.length} onChange={this.handleChange} type="number" min="1" required/>
        </label>

        <h2>Options</h2>

        <Tabs>
          <TabList>{this.constructor.tabs}</TabList>
          {this.tabPanels}
        </Tabs>

        <h2>Possible Passwords</h2>
        {approximatePrefix}{possiblePasswords.toLocaleString()} ({approximatePrefix}{Math.log2(possiblePasswords).toFixed(2)} bits of entropy)
      </form>
    )
  }
}
