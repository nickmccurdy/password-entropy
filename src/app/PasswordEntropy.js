import CustomOptions from '../options/CustomOptions'
import DicewareOptions from '../options/DicewareOptions'
import FormGroup from '../ui/FormGroup'
import GenericOptions from '../options/GenericOptions'
import PossiblePasswords from './PossiblePasswords'
import PropTypes from 'prop-types'
import { pick } from 'ramda'
import React, { PureComponent } from 'react'
import { Form, PageHeader } from 'react-bootstrap'
import Icon from 'react-fa'
import { connect } from 'react-redux'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs'
import 'react-tabs/style/react-tabs.css'

export class PasswordEntropy extends PureComponent {
  static options = [
    {
      Component: DicewareOptions,
      icon: 'book'
    },
    {
      Component: CustomOptions,
      icon: 'question-circle'
    },
    {
      Component: GenericOptions,
      icon: 'check-square'
    }
  ]

  static propTypes = {
    length: PropTypes.number.isRequired,
    onChange: PropTypes.func.isRequired
  }

  render () {
    return (
      <div>
        <PageHeader>
          <Icon name="lock"/> Password Entropy
          <br/>
          <small>A simple entropy calculator for evaluating password security.</small>
        </PageHeader>


        <Form horizontal>
          <FormGroup id="length" label="Length" icon="arrows-h">
            <input name="length" value={this.props.length} onChange={this.props.onChange} type="number" min="1" required/>
          </FormGroup>

          <h2><Icon name="cog"/> Options</h2>

          <Tabs>
            <TabList>
              {this.constructor.options.map(({ Component: { shortName }, icon }) => (
                <Tab key={shortName}><Icon name={icon}/> {shortName}</Tab>
              ))}
            </TabList>
            {this.constructor.options.map(({ Component }) => (
              <TabPanel key={Component.shortName}>
                <Component/>
              </TabPanel>
            ))}
          </Tabs>

          <FormGroup id="results" label="Results" icon="info-circle">
            <PossiblePasswords length={this.props.length}/>
          </FormGroup>
        </Form>
      </div>
    )
  }
}

export const mapStateToProps = pick(['length'])

export const mapDispatchToProps = {
  onChange: length => ({
    type: 'SET_LENGTH',
    payload: parseInt(length, 10)
  })
}

export default connect(mapStateToProps, mapDispatchToProps)(PasswordEntropy)
