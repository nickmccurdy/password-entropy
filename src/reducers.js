import GenericOptions from './options/GenericOptions'
import { assoc, identity, map, T } from 'ramda'
import { combineReducers } from 'redux'

function createReducer(actionType, initialState, { callback = identity, validator } = {}) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = !validator || validator(payload)

    return matchesAction && isValid ? callback(payload, state) : state
  }
}

export default combineReducers({
  length: createReducer('SET_LENGTH', 6, { validator: n => n >= 1 }),
  options: combineReducers({
    custom: createReducer('SET_CUSTOM', 1, { validator: n => n >= 0 }),
    generic: createReducer ('TOGGLE_GENERIC', map(T, GenericOptions.toggles), {
      callback: ({ checked, name }, state) => assoc(name, checked, state)
    })
  }),
  optionsKey: createReducer('SET_OPTIONS_KEY', 'diceware')
})
