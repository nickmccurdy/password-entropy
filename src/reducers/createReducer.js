import { identity } from 'ramda'

export default function createReducer(actionType, initialState, callback = identity) {
  return (state = initialState, { payload, type }) => {
    const matchesAction = type === actionType
    const isValid = typeof payload !== 'number' || payload >= 0

    return matchesAction && isValid ? callback(payload, state) : state
  }
}
