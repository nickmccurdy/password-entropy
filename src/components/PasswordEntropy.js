import arrowsAltH from '@fortawesome/fontawesome-free-solid/faArrowsAltH'
import infoCircle from '@fortawesome/fontawesome-free-solid/faInfoCircle'
import lock from '@fortawesome/fontawesome-free-solid/faLock'
import Icon from '@fortawesome/react-fontawesome'
import PropTypes from 'prop-types'
import React from 'react'
import FormControl from 'react-bootstrap/lib/FormControl'
import PageHeader from 'react-bootstrap/lib/PageHeader'
import { connect } from 'react-redux'

import FormGroup from './FormGroup'
import { selector, setLength } from '../reducers/length'
import Options from './Options'
import PossiblePasswords from './PossiblePasswords'

export function PasswordEntropy({ length, setLength }) {
  return (
    <div>
      <PageHeader>
        <Icon icon={lock} /> Password Entropy
        <br />
        <small>A simple entropy calculator for evaluating password security.</small>
      </PageHeader>

      <form>
        <FormGroup label="Results" icon={infoCircle}>
          <PossiblePasswords />
        </FormGroup>

        <FormGroup label="Length" icon={arrowsAltH}>
          <FormControl value={length} onChange={setLength} type="number" min="0" required />
        </FormGroup>

        <Options />
      </form>
    </div>
  )
}

PasswordEntropy.propTypes = {
  length: PropTypes.number.isRequired,
  setLength: PropTypes.func.isRequired
}

export default connect(selector, { setLength })(PasswordEntropy)
