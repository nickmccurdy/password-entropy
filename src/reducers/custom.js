import { path } from 'ramda'

import createReducer from './createReducer'

export default createReducer('SET_CUSTOM', 0)

export function setCustom({ target: { value } }) {
  return { type: 'SET_CUSTOM', payload: parseInt(value, 10) }
}

export const selector = state => ({
  possibleItems: path(['options', 'custom'], state)
})
