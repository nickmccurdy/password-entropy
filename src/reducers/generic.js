import assoc from 'ramda/src/assoc'
import map from 'ramda/src/map'
import path from 'ramda/src/path'
import T from 'ramda/src/T'

import createReducer from './createReducer'

export const toggles = {
  Letters: {
    example: 'a-z',
    possibleItems: 26
  },
  'Capital Letters': {
    example: 'A-Z',
    possibleItems: 26
  },
  Numbers: {
    example: '0-9',
    possibleItems: 10
  },
  Symbols: {
    example: '!@#$%^&*',
    possibleItems: 8
  }
}

export const defaultToggles = map(T, toggles)

export function toggleGenericCallback({ checked, name }, state) {
  return assoc(name, checked, state)
}

export default createReducer(
  'TOGGLE_GENERIC',
  defaultToggles,
  toggleGenericCallback
)

export function toggleGeneric({ target: { name } }, checked) {
  return {
    type: 'TOGGLE_GENERIC',
    payload: {
      checked,
      name
    }
  }
}

export const selector = path(['options', 'Generic'])
