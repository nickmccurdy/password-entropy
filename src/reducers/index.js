import { combineReducers } from 'redux'

import custom from './custom'
import generic from './generic'
import length from './length'
import optionsKey from './optionsKey'

export default combineReducers({
  length,
  options: combineReducers({ custom, generic }),
  optionsKey
})
