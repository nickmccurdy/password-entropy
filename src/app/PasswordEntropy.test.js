import { shallow } from 'enzyme'
import { PasswordEntropy, mapStateToProps, mapDispatchToProps } from './PasswordEntropy'
import React from 'react'
import reducers from '../reducers'
import { createStore } from 'redux'

test('capitalize', () => {
  expect(PasswordEntropy.capitalize('word')).toEqual('Word')
})

it('renders', () => {
  const wrapper = shallow(<PasswordEntropy length={6} onChange={() => {}} onSelect={() => {}} optionsKey="diceware"/>)
  expect(wrapper).toMatchSnapshot()
})

test('mapStateToProps', () => {
  const state = createStore(reducers).getState()
  expect(mapStateToProps(state)).toEqual({ length: 6, optionsKey: 'diceware' })
})

test('mapDispatchToProps', () => {
  expect(mapDispatchToProps.onChange({ target: { value: '6' } })).toEqual({
    type: 'SET_LENGTH',
    payload: 6
  })

  expect(mapDispatchToProps.onSelect(0)).toEqual({
    type: 'SET_OPTIONS_KEY',
    payload: 0
  })
})
