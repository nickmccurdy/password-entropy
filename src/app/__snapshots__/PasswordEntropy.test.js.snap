// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordEntropy 1`] = `
<div>
  <PageHeader
    bsClass="page-header"
  >
    <Icon
      Component="span"
      name="lock"
    />
     Password Entropy
    <br />
    <small>
      A simple entropy calculator for evaluating password security.
    </small>
  </PageHeader>
  <form>
    <FormGroup
      icon="info-circle"
      id="results"
      label="Results"
    >
      <Connect(PossiblePasswords) />
    </FormGroup>
    <FormGroup
      icon="arrows-h"
      id="length"
      label="Length"
    >
      <FormControl
        bsClass="form-control"
        componentClass="input"
        min="0"
        onChange={[Function]}
        required={true}
        type="number"
        value={12}
      />
    </FormGroup>
    <Connect(Options) />
  </form>
</div>
`;
